### PR TITLE
PT-264/Add support for multiple dashboard instances

### PR DESCRIPTION
**Pull request by florianmski**
Tracked in JIRA by [PT-342](https://novoda.atlassian.net/browse/PT-342)

### Scope of the PR
This PR is the first part to have different dashboards on different screens.
This just launches different instances of the `Dashboard` whenever we use a different id in the url `.../dashboards/:id`.
Next PR will be about having different configs for different instances.

Paired with @eduardb

### CODE REVIEW

File: angular2-client/src/app/dashboard-carousel/dashboard-carousel.component.ts [hunk 1]
```diff
@@ -30,30 +31,42 @@ const COMPONENTS = {
...
   ngOnInit(): void {
+    this.routeParams.params.subscribe(params => {
+        this.id = params['id'];
+    });
     this.configSubscription = this.configService.getServerUrl()
-      .map((serverUrl: string) => {
-        return this.socketService.create(serverUrl);
+      .zip(this.routeParams.params, this.createServerConfig)
+      .map(serverConfig => {
+        return this.socketService.create(serverConfig.url, serverConfig.id);
```
Comment (florianmski) [inline angular2-client/src/app/dashboard-carousel/dashboard-carousel.component.ts]: We now send the requested id for the dashboard to the server.
AS in `.../dashboards/london` will send the `london` id

File: node-server/index.js [hunk 2]
```diff
@@ -1,37 +1,11 @@
...
-
-app.get('/', function(req, res) {
-  res.sendFile(__dirname + '/client/index.html');
-});
-
-app.get('/refresh', function(req, res) {
-  res.sendStatus(200);
-  io.emit('refresh');
-});
-
```
Comment (florianmski) [inline node-server/index.js]: :x: This was needed for the old client, we don't need it anymore

File: node-server/server/dashboard.js [hunk 3]
```diff
@@ -8,10 +8,11 @@ const PeopleWall = require('./widget/people-wall');
 const Calendar = require('./widget/calendar');
 const GithubContributors = require('./widget/github-reports/github-contributors');
 
-const DASHBOARD_INTERVAL = 1000 * 30;
+const DASHBOARD_INTERVAL = 1000 * 5;
```
Comment (florianmski) [inline node-server/server/dashboard.js]: This was only for testing, need to be reverted

File: node-server/server/dashboard.js [hunk 4]
```diff
@@ -26,21 +27,27 @@ function Dashboard(config) {
...
-    getCurrentRule(self).then(result => {
-        listener(result)
-        incrementIndex(self);
-        setTimeout(updateLoop(self, listener), getTimeoutInterval(self));
-      }).catch(err => {
+    getCurrentRule(self)
+      .then(result => {
+        if (self.callback) {
+          self.callback(self, result);
+          self.lastResult = result;
```
Comment (florianmski) [inline node-server/server/dashboard.js]: We are now caching the result so we can send it to new connected client directly

File: node-server/server/dashboard.js [hunk 5]
```diff
@@ -8,10 +8,11 @@ const PeopleWall = require('./widget/people-wall');
...
 const GithubContributors = require('./widget/github-reports/github-contributors');
 
-const DASHBOARD_INTERVAL = 1000 * 30;
+const DASHBOARD_INTERVAL = 1000 * 5;
 const DASHBOARD_ERROR_INTERVAL = 1000 * 1;
 
-function Dashboard(config) {
+function Dashboard(id, config) {
+  this.id = id;
   this.slacker = new Slacker(config.slack);
```
Comment (florianmski) [inline node-server/server/dashboard.js]: :x: We need to have only one instance of this

File: angular2-client/src/app/app.routes.ts [hunk 6]
```diff
@@ -3,7 +3,7 @@ import { PageNotFoundComponent } from './page-not-found/page-not-found.component
 import { DashboardCarouselComponent } from './dashboard-carousel/dashboard-carousel.component';
 
 const appRoutes: Routes = [
-  {path: 'dashboards', component: DashboardCarouselComponent},
+  {path: 'dashboards/:id', component: DashboardCarouselComponent},
```
Comment (frapontillo) [inline angular2-client/src/app/app.routes.ts]: 👍

File: angular2-client/src/app/dashboard-carousel/socket.service.ts [hunk 7]
```diff
@@ -7,18 +7,19 @@ export class SocketService {
 
   private observable: ConnectableObservable<MessageEvent>;
 
-  public create(url): ConnectableObservable<MessageEvent> {
+  public create(url: string, id: string): ConnectableObservable<MessageEvent> {
```
Comment (frapontillo) [inline angular2-client/src/app/dashboard-carousel/socket.service.ts]: Maybe call this `roomId`?

File: angular2-client/src/app/dashboard-carousel/socket.service.ts [hunk 8]
```diff
@@ -7,18 +7,19 @@ export class SocketService {
...
     if (!this.observable) {
-      this.observable = this.createObservable(url);
+      this.observable = this.createObservable(url, id);
     }
     return this.observable;
   }
 
-  private createObservable(url): ConnectableObservable<any> {
+  private createObservable(url: string, id: string): ConnectableObservable<any> {
```
Comment (frapontillo) [inline angular2-client/src/app/dashboard-carousel/socket.service.ts]: Same here please :D

File: node-server/server/clients.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
+function Clients() {
```
Comment (frapontillo) [inline node-server/server/clients.js]: But this is ES6 on Node, you have `class` and should therefore avoid prototypes.

File: node-server/server/clients.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+Clients.prototype.remove = function(client) {
+  let position = this.allClients.indexOf(client);
+  if (position != -1) {
+    this.allClients.splice(position, 1);
+  }
+}
+
+Clients.prototype.forDashboardId = function(dashboardId) {
+  return this.allClients.filter(isClientOf.bind(this, dashboardId));
```
Comment (frapontillo) [inline node-server/server/clients.js]: Having classes will remove the need for such trickery 😉

File: node-server/server/clients.spec.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  let clients;
+  let dashboardId = 0;
+  let client = {handshake: {query: {id: dashboardId}}};
+
+  beforeEach(function() {
+    clients = new Clients();
+  });
+
+  it('should hold client when calling put', () => {
+
```
Comment (frapontillo) [inline node-server/server/clients.spec.js]: Blank line 🚓

File: node-server/server/dashboard.js [hunk 12]
```diff
@@ -26,21 +27,27 @@ function Dashboard(config) {
   ].concat(this.slacker.getRules());
 }
 
-Dashboard.prototype.start = function(listener) {
-  updateLoop(this, listener)();
+Dashboard.prototype.start = function(callback) {
```
Comment (frapontillo) [inline node-server/server/dashboard.js]: Can be reimplemented as a class.

File: node-server/server/dashboards.js [hunk 13]
```diff
@@ -0,0 +1,26 @@
+const Dashboard = require('./dashboard.js');
+
+function Dashboards() {
```
Comment (frapontillo) [inline node-server/server/dashboards.js]: Can be implemented as a class.

**Comment by florianmski** [comment]
Still need to fix some :x: in my comments, let's allow only the label setter to remove the label

File: angular2-client/src/app/dashboard-carousel/socket.service.ts (hunk 7 above)
Comment (frapontillo) [inline angular2-client/src/app/dashboard-carousel/socket.service.ts]: Makes sense! 👍

File: node-server/server/clients.js [hunk 14]
```diff
@@ -1,28 +1,33 @@
...
+  forDashboardId(dashboardId) {
+    return this.allClients.filter(client => this._isClientOf(this, dashboardId, client));
+  }
+
+  dashboardIdFromClient(client) {
+    return client.handshake.query.id;
+  }
+
+  _isClientOf(clients, dashboardId, client) {
+     return clients.dashboardIdFromClient(client) == dashboardId;
```
Comment (frapontillo) [inline node-server/server/clients.js]: Use `===` 😄

File: node-server/server/dashboards.js [hunk 15]
```diff
@@ -1,26 +1,30 @@
...
+    }
+  }
 
-Dashboards.prototype.stop = function(dashboard) {
-  let position = this.allDashboards.indexOf(dashboard);
-  if (position != -1) {
-    dashboard.stop();
-    this.allDashboards.splice(position, 1);
+  instanceFor(id) {
+    return this.allDashboards.find(dashboard => dashboard.id == id);
```
Comment (frapontillo) [inline node-server/server/dashboards.js]: Same `===` since we're at it

File: node-server/server/dashboards.js [hunk 16]
```diff
@@ -1,26 +1,30 @@
...
   }
+
+  stop(dashboard) {
+    let position = this.allDashboards.indexOf(dashboard);
+    if (position != -1) {
```
Comment (frapontillo) [inline node-server/server/dashboards.js]: And `!==`

File: node-server/server/clients.spec.js (hunk 11 above)
Comment (eduardb) [inline node-server/server/clients.spec.js]: Yeah, that's to show that we don't have any `given`

Comment (frapontillo) [inline node-server/server/clients.spec.js]: It's not usually a thing in JS, but I'll concede!

**Comment by florianmski** [comment]
That's good to know 👍  @frapontillo

**Comment by xrigau** [comment]
💯 cool stuff!! 🎉

**Pull request merged**